### PR TITLE
Validate by github action if phpunit tests don't change repos state

### PR DESCRIPTION
**Pull request by MeKeyCool**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | When i run `php -d date.timezone=UTC ./vendor/phpunit/phpunit/phpunit -c tests/Unit/phpunit.xml tests/Unit/Core/Module` some of test resources are deleted.
| Type?             | bug fix
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Na.
| Related PRs       | 
| How to test?      | CI should pass; run `php -d date.timezone=UTC ./vendor/phpunit/phpunit/phpunit -c tests/Unit/phpunit.xml tests/Unit/Core/Module` command, then `git status` nothing should have changed.
| Possible impacts? | CI and tests could be broken.

### CODE REVIEW

File: tests/Unit/Core/Module/SourceHandler/ZipSourceHandlerTest.php
Comment (matks) [inline tests/Unit/Core/Module/SourceHandler/ZipSourceHandlerTest.php]: It is not a directory anymore 😛 so the constant name must be changed or the logic must be changed

File: .github/workflows/php.yml
Comment (matks) [inline .github/workflows/php.yml]: Why put this in the same githubaction step? Should'nt we make it a specific step?

**Review by matks — CHANGES_REQUESTED** [review]

File: .github/workflows/php.yml
Comment (MeKeyCool) [reply .github/workflows/php.yml]: As it is phpunit that remove some files I thought it is a part of this Action to ensure the state is the same at the beginning and at the end.

What do you recommend ? 
Should I run two distinct Actions with 
```yaml
         run: ./vendor/phpunit/phpunit/phpunit -c tests/Unit/phpunit.xml
```
and
```yaml
         run: ./vendor/phpunit/phpunit/phpunit -c tests/Unit/phpunit.xml; git diff --exit-code
```
?

Or maybe we can assume that an Action is dependent of preceding Action ? :thinking:

File: tests/Unit/Core/Module/SourceHandler/ZipSourceHandlerTest.php
Comment (MeKeyCool) [reply tests/Unit/Core/Module/SourceHandler/ZipSourceHandlerTest.php]: Why do you say it is not a directory ? :thinking: 
Maybe you don't like `.tmp` on a directory name ? Would you prefer `zip_source_handler_tmp` ? Or `tmp_zip_source_handler` ?

Comment (MeKeyCool) [reply tests/Unit/Core/Module/SourceHandler/ZipSourceHandlerTest.php]: Or Maybe `zip_source_handler_test`

**Review by jolelievre — COMMENTED** [review]
Hi @MeKeyCool 

I noticed this problem as well, it should be fixed by this PR https://github.com/PrestaShop/PrestaShop/pull/27055

I added a method in our `ResourceResetter` class to restre the content of the modules test directory. So the test that mess with are responsible for cleaning/restoring the initial state when they ended by using `(new ResourceResetter())->resetTestModules();`.

I think it's cleaner if each test class cleans the context rather than the CI itself so that even in local everything is clean. I didn't integrate it in `ZipSourceHandlerTest` though because I didn't notice this one needed some cleaning as well.

Maybe it should be improved as well once my PR is merged by using the same method.

**Review by jolelievre — COMMENTED** [review]
Hi @MeKeyCool 

I noticed this problem as well, it should be fixed by this PR https://github.com/PrestaShop/PrestaShop/pull/27055

I added a method in our `ResourceResetter` class to restre the content of the modules test directory. So the test that mess with are responsible for cleaning/restoring the initial state when they ended by using `(new ResourceResetter())->resetTestModules();`.

I think it's cleaner if each test class cleans the context rather than the CI itself so that even in local everything is clean. I didn't integrate it in `ZipSourceHandlerTest` though because I didn't notice this one needed some cleaning as well.

Maybe it should be improved as well once my PR is merged by using the same method.

**Comment by MeKeyCool** [comment]
> Hi @MeKeyCool
> 
> I noticed this problem as well, it should be fixed by this PR #27055
> 
> I added a method in our `ResourceResetter` class to restre the content of the modules test directory. So the test that mess with are responsible for cleaning/restoring the initial state when they ended by using `(new ResourceResetter())->resetTestModules();`.
> 
> I think it's cleaner if each test class cleans the context rather than the CI itself so that even in local everything is clean. I didn't integrate it in `ZipSourceHandlerTest` though because I didn't notice this one needed some cleaning as well.
> 
> Maybe it should be improved as well once my PR is merged by using the same method.

Sounds good, can you add WIP flag to this PR and I'll review it once your PR will be merged ^^

**Comment by MeKeyCool** [comment]
Please remove the "blocked" tag :slightly_smiling_face: 
This PR is ready for review ^^

File: .github/workflows/php.yml
Comment (Progi1984) [inline .github/workflows/php.yml]:
```suggestion
        run: ./vendor/phpunit/phpunit/phpunit -c tests/Unit/phpunit.xml
```

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Could you fix the CI, please ?

File: .github/workflows/php.yml
Comment (MeKeyCool) [reply .github/workflows/php.yml]: No question ? No argue ? ^^' ...
I didn't make this change like a random bot xD

:thinking: Hmm ... Should I take it like an order ? ^^' 

Then I changed this cause phpunit command changed some git versioned files and this behavior wasn't detected by the CI but induced some mess in local environment after running tests.

Do you have any suggestion to ensure that tests won't break anything in environment ? :slightly_smiling_face:

**Comment by MeKeyCool** [comment]
> Could you fix the CI, please ?

Sorry for that ^^'

File: .github/workflows/php.yml
Comment (Progi1984) [reply .github/workflows/php.yml]: You can add a final test in Github Action where to check that the output is empty.

No chain commands, please.

Comment (MeKeyCool) [reply .github/workflows/php.yml]: Thanks for explanations !! ^^

**Comment by MeKeyCool** [comment]
As i received several comments about my change on github action, I created this temporary PR to explain more precisely : https://github.com/PrestaShop/PrestaShop/pull/28413

In case a developer doesn't run tests before to open a PR, this CI test tests phpunit tests and checks if they don't change anything in versioned files state.
This should ensure more comfortable environment for developers and help to not commit some changes by error.

**Comment by matks** [comment]
@MeKeyCool 

> In case a developer doesn't run tests before to open a PR, this CI test tests phpunit tests and checks if they don't change anything in versioned files state.

In other words, your goal is
1. I run `git status` : nothing reported not tracked or changed
2. I run CI
3. I run `git status` : I should have the same output, no untracked/changed file

Right?

**Review by matks — DISMISSED** [review]

**Comment by MeKeyCool** [comment]
> In other words, your goal is
> 
>     1. I run `git status` : nothing reported not tracked or changed
> 
>     2. I run CI
> 
>     3. I run `git status` : I should have the same output, no untracked/changed file
> 
> 
> Right?

Not exactly,

My goal with this test is more to ensure that, in my ***development environment***
1. I `git clone`/`git pull`/`git rebase` from official Prestashop/Prestashop clean sources.
1. I run `composer unit-test`
1. I run `git status` : nothing reported not tracked or changed

And the proposed strategy is, in ***CI environment***
1. Run `composer unit-test`
2. Check if `git status` have no untracked/changed file

This way, any test added or changed that induce some untracked/changed files in repository should be rejected by CI.

File: .github/workflows/php.yml
Comment (atomiix) [inline .github/workflows/php.yml:127]: I guess we could extend this to integration tests and behavior tests too?

**Review by atomiix — APPROVED** [review]

**Review by matks — APPROVED** [review]
Good for me.

I would like a 3rd opinion @PrestaShop/prestashop-maintainers

**Review by Progi1984 — APPROVED** [review]

**Comment by Progi1984** [comment]
Thanks @MeKeyCool

**Pull request merged**
